fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -0,0 +1,1839 @@
+	loop {
+		wait(delay(SERVER_KNOBS->ROCKSDB_METRICS_DELAY));
+		TraceEvent e("RocksDBFlowLock");
+		e.detail("ReadAvailable", readLock->available());
+		e.detail("ReadActivePermits", readLock->activePermits());
+		e.detail("ReadWaiters", readLock->waiters());
+		e.detail("FetchAvailable", fetchLock->available());
+		e.detail("FetchActivePermits", fetchLock->activePermits());
+		e.detail("FetchWaiters", fetchLock->waiters());
+	}
+}
+
+struct PhysicalShard;
+
+struct DataShard {
+	DataShard(KeyRange range, PhysicalShard* physicalShard) : range(range), physicalShard(physicalShard) {}
+
+	KeyRange range;
+	PhysicalShard* physicalShard;
+};
+
+struct PhysicalShard {
+	PhysicalShard(rocksdb::DB* db, std::string id) : db(db), id(id) {}
+
+	void init() {
+		if (cf) {
+			return;
+		}
+		cf = createNewColumnFamily(db, id);
+		readIterPool = std::make_shared<ReadIteratorPool>(db, cf, id);
+	}
+
+	~PhysicalShard() {
+		readIterPool.reset();
+		if (!deletePending)
+			return;
+
+		// Destroy CF
+		auto s = db->DropColumnFamily(cf);
+		if (!s.ok()) {
+			logRocksDBError(s, "DestroyShard");
+			logShardEvent(id, ShardOp::DESTROY, SevError, s.ToString());
+			return;
+		}
+		logShardEvent(id, ShardOp::DESTROY);
+	}
+
+	rocksdb::DB* db;
+	std::string id;
+	rocksdb::ColumnFamilyHandle* cf;
+	std::unordered_map<std::string, std::unique_ptr<DataShard>> dataShards;
+	std::shared_ptr<ReadIteratorPool> readIterPool;
+	bool deletePending = false;
+};
+
+class ShardManager {
+public:
+	ShardManager(std::string path) : path(path) {}
+	rocksdb::Status init() {
+		// TODO: Open default instance and get shard mapping.
+		// Open Instances
+		auto options = getOptions();
+		rocksdb::Status status = rocksdb::DB::Open(options, path, &db);
+		if (!status.ok()) {
+			logRocksDBError(status, "Open");
+			return status;
+		}
+		metadataShard = std::make_shared<PhysicalShard>(db, "kvs-metadata");
+		physicalShards["kvs-metadata"] = metadataShard;
+		metadataShard->init();
+		return status;
+	}
+
+	void closeAllShards() {
+		// Close DB.
+		auto s = db->Close();
+		if (!s.ok()) {
+			logRocksDBError(s, "Close");
+			return;
+		}
+	}
+
+	void destroyAllShards() {
+		closeAllShards();
+		std::vector<rocksdb::ColumnFamilyDescriptor> cfs;
+		for (const auto& [key, _] : physicalShards) {
+			cfs.push_back(rocksdb::ColumnFamilyDescriptor{ key, getCFOptions() });
+		}
+		auto s = rocksdb::DestroyDB(path, getOptions(), cfs);
+		if (!s.ok()) {
+			logRocksDBError(s, "DestroyDB");
+		}
+		TraceEvent("RocksDB").detail("Info", "DBDestroyed");
+	}
+
+	rocksdb::DB* getDb() { return db; }
+
+private:
+	std::string path;
+	rocksdb::DB* db = nullptr;
+	std::unordered_map<std::string, std::shared_ptr<PhysicalShard>> physicalShards;
+	std::unique_ptr<rocksdb::WriteBatch> writeBatch;
+	KeyRangeMap<DataShard*> dataShardMap;
+	std::shared_ptr<PhysicalShard> metadataShard = nullptr;
+};
+
+int readRangeInDb(DataShard* shard, const KeyRangeRef& range, int rowLimit, int byteLimit, RangeResult* result) {
```

Comment:
nit: can `readRangeInDb` be a member function of `DataShard`?

Reply:
I'll update it when sending out PR for read range.